管理者 <-> サイトの関連づけ [close #198]

mtsmfm opened the pull request:
- editor can do what under /editor
- admin can do what under /admin

mtsmfm commented on app/controllers/admin/application_controller.rb:
```diff
@@ -2,4 +2,33 @@ class Admin::ApplicationController < ApplicationController
+  def can?(action, resource)
+    case resource
+    when Class
+      can?(action, resource.new)
```
素直になんか gem 使うべきな気がする...

tricknotes commented:
サイトオーナーと管理者では管理画面を開くときのメンタルモデルが違うように思えたので、まるっと画面をわけた方が進めやすいのかなと感じました。
サイトオーナーは現在表示しているサイト(とそれに紐づくリソース)に興味があるけど、管理者はユーザ一覧とサイト一覧が見えていれば十分かと。
(そうすると、権限管理もあまり複雑な仕組みにしなくても十分に思います)

okkez commented:
:+1: いい気がします

tricknotes commented:
サイトオーナーが User <-> Site の関連のみで表現されるのであれば、`role: 'site_owner'` は必要なかったりしますか？

mtsmfm commented on db/migrate/20160201092922_add_role_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :role, :string, null: false, default: 'site_owner'
```
`admin` の boolean だけでよさそう

mtsmfm commented:
> サイトオーナーが User <-> Site の関連のみで表現されるのであれば、role: 'site_owner' は必要なかったりしますか？

exactly!

mtsmfm committed:
Add memberships and admin flag

mtsmfm committed:
Pretty admin/editor page

- Separate admin | editor area
- Add user management for admin
- Add some tests

tricknotes commented:
active_link_to が意図しない active をつけている、ような。
![2016-02-04 18 50 44](https://cloud.githubusercontent.com/assets/290782/12811510/415df548-cb70-11e5-88eb-0537c94c255a.png)

mtsmfm commented:
> active_link_to が意図しない active をつけている、ような。

:see_no_evil:

mtsmfm commented on config/routes.rb:
```diff
@@ -6,13 +6,21 @@
 
-    resource  :site, only: %i(edit update)
+    resources :sites, except: :destroy do
+      resources :fixed_pages
```
site の持ち物は editor が管理するようにしたほうが素直では〜という話をしました

mtsmfm committed:
Move sites' resources to editor

mtsmfm committed:
Remove unnecessary visit

tricknotes commented:
自分のパスワード変更ができないですが、これはまた今度やりますか？(今までもできなかったですし)

tricknotes commented on app/views/editor/application/_navigation.html.haml:
```diff
@@ -0,0 +1,14 @@
+        = active_link_to 'カテゴリ', editor_categories_path, wrap_tag: 'li'
+        = active_link_to '記事',     editor_posts_path,      wrap_tag: 'li'
+        = active_link_to '固定ページ', editor_fixed_pages_url, wrap_tag: 'li'
+        = active_link_to 'リンク', editor_links_url(site_id: current_site), wrap_tag: 'li'
```
Is `site_id` needed?

mtsmfm replied:
:dizzy:

tricknotes commented on app/views/editor/application/_navigation.html.haml:
```diff
@@ -0,0 +1,14 @@
+      %ul.nav.navbar-nav.navbar-right
+        = active_link_to 'カテゴリ', editor_categories_path, wrap_tag: 'li'
+        = active_link_to '記事',     editor_posts_path,      wrap_tag: 'li'
+        = active_link_to '固定ページ', editor_fixed_pages_url, wrap_tag: 'li'
```
Which do you want to use `xxx_path` / `xxx_url` ?

mtsmfm replied:
In some situations `_path` doesn't work well (ex. CORS).
But `_url` always works.
So I'll replace `_path` with `_url`

tricknotes replied:
Reasonable :+1:

mtsmfm replied:
By the way, this PR has many diffs so I'll fix url helper on another PR :wink:

tricknotes replied:
make sense :-)

tricknotes commented on app/views/layouts/admin.html.haml:
```diff
@@ -2,13 +2,13 @@
 
   %body
     .container
-      = render 'admin_navigation'
```
This partial might be unreferenced from anywhere.

mtsmfm replied:
good catch :baseball:

tricknotes commented on app/views/editor/application/_navigation.html.haml:
```diff
@@ -0,0 +1,14 @@
+    .container-fluid
+      .navbar-header
+        .navbar-brand "#{current_site.name}"
+
```
Is it intentional what the switching link to other site and the logo image has gone?
(It is current `admin_navigation`s behavior.)

mtsmfm replied:
I'll restore current behavior

tricknotes commented:
I left some comments.
However this seems an awesome work :sparkles:

okkez commented:
Looks good to me.
Great work! :sparkler:

okkez commented:
Please write about current architecture what is  in merge comment when merge this PR.

* editor can do what under `/editor`
* admin can do what under `/admin`

mtsmfm committed:
Remove unnecessary param

mtsmfm committed:
Restore dropdown links

mtsmfm committed:
Remove unused partial

tricknotes commented:
just awesome :whale:
